### PR TITLE
Fixes #27160 - do not use Ruby Timeout for DNS

### PR DESCRIPTION
**Pull request by lzap**
We use Timeout Ruby class in our codebase to handle DNS queries which is not good approach. Actually Ruby DNS Resolver reads all "nameserver" entries and tries one after another, however the way this is implemented is that the default Ruby DNS timeout is high (5, 10, 20, 40 seconds) and it probably does not reply fast enough and the global timeout (in our codebase) fires off.

The solution is to remove the extra timeout we have there and rely only on the Ruby DNS resolver timeout. No idea why this wasn't there from the day one. Try the patch and let me know if it solved the problem.

This issue was reported by a community user.

### CODE REVIEW

**Comment by ananace** [comment]
Tested the patch, seems to solve the issue, though it does lead to a not insubstantial delay in actions that cause lookups. Which is to be expected I suppose.

:+1:

**Comment by lzap** [comment]
Thanks, well depends on what you have in the setting, I think it's 3 seconds by default which is not bad. So default behavior for two nameservers A and B should be:

* query A
* wait 3 seconds, timeout
* query B
* wait 3 seconds, timeout
* raise timeout exception

**Comment by lzap** [comment]
https://github.com/ruby/ruby/blob/trunk/lib/resolv.rb#L1117-L1129 FYI

File: lib/net/dns.rb
Comment (lzap) [inline lib/net/dns.rb:56]: No changes in the codebase only reindented because of the timeout block not needed.

File: lib/net/dns.rb
Comment (lzap) [inline lib/net/dns.rb:62]: This exception is not longer thrown, DNS resolver issues `Resolv::ResolvTimeout` in case of timeout which is passed over the callstack. Which reminds me to checck if we rescue `Net::Error` because this changes behavior - and we do! Rebased.

**Comment by lzap** [comment]
Foreman gree, Katello test passed the last time, probably a CI issue. [test katello]

File: lib/foreman/renderer/scope/macros/base.rb
Comment (ekohl) [inline lib/foreman/renderer/scope/macros/base.rb:69]: TIL we have a DNS lookup macro. Looks like nothing uses this so I wonder if we should deprecate it. Doesn't have to impact this PR.

**Review by ekohl — COMMENTED** [review]
I would like to kill all DNS lookups from Foreman. The Proxy can do conflict resolution and otherwise we shouldn't care. I can dream, right? :)

This looks correct but I haven't tested it.

File: lib/foreman/renderer/scope/macros/base.rb
Comment (lzap) [reply lib/foreman/renderer/scope/macros/base.rb:69]: Let's keep it, this was requested macro by a RH customer: https://projects.theforeman.org/issues/16057

**Comment by lzap** [comment]
> I would like to kill all DNS lookups from Foreman. The Proxy can do conflict resolution and otherwise we shouldn't care. I can dream, right? :)

There are couple of other places we do DNS lookups for, like reverse lookup of smart-proxy hostname to acquire TFTP IP address. It can also be freely used in templates. But I agree we should not probably do lookups during DNS orchestration that's quite confusing.

**Comment by lzap** [comment]
Ping, or maybe @ares with this one - I would like to merge this one before the tftp_servername item from my BZ plate.

**Comment by ekohl** [comment]
This does remove the retry in case a DNS query fails. I find it hard to imagine the exact impact of this.

**Comment by lzap** [comment]
Rebased, it's now an array. Proper migration implemented including rollback, tested.

I am suggesting the follofing default value: (5, 10, 15, 20) which is 50 seconds in total. I intentionally made it less than a minute because 60 seconds is often a timeout (e.g. http timeout) so there is some "room" for the process to finish. Ruby has more complex logic in calculating the default value, but it is pretty close.

File: app/models/setting/provisioning.rb
Comment (ekohl) [inline app/models/setting/provisioning.rb]: Can we rename this to `dns_timeout`? It's not just for conflict checking.

Comment (lzap) [reply app/models/setting/provisioning.rb]: Aaaaaghhhhh then I cannot probably use migration because Rails will initialize this. Damn. Need to start over.

Comment (ekohl) [reply app/models/setting/provisioning.rb]: I think it could be ok not to have a migration. This is a very different implementation and we can warn users about this in the upgrade notes.

**Comment by lzap** [comment]
Ok brand new migration which handles renaming properly and also migrates the value (if it was set to something different than the funky default 3 seconds).

**Comment by lzap** [comment]
Katello error irrelevant `redirection_url_on_host_deletion` and `copy_all_units_test`.

**Comment by ares** [comment]
I haven't tested yet, but I think this version should make everyone happy :-) I'll wait for confirmation before moving on with testing.

File: db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb
Comment (ekohl) [inline db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: Do we want to set this to `[old_setting.value / 2, old_setting.value / 2]` so there's at least one retry?

**Review by ekohl — COMMENTED** [review]
Untested, but overall :+1:. I very much like that the setting name is now much more accurate even though this should be mentioned in https://github.com/theforeman/theforeman.org/blob/gh-pages/_includes/manuals/nightly/1.2_release_notes.md#upgrade-warnings after merging.

File: db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb
Comment (lzap) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: I was thinking the same, then I preferred simplicity. It worked so it will probably work too. Neverless, I believe that majority of users never touched this setting - it's good as a workaround enabler.

Comment (lzap) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: Can do another rebase if you insist.

Comment (lzap) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: Any chance to get it in? I have it as a sprint task, many thanks!

Comment (ekohl) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: I don't care that much (as I said, :+1:). This can also be mentioned in the upgrade warnings. When the user changed the previous setting, they should always look at this new setting after upgrading. If you can mention that in an upgrade warning, then I'm fine with merging. I think @ares wants to test this so I'll leave the honors to him.

Comment (lzap) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: I don't understand what do I need to warn about - the upgrade is seamless. Foreman will respect what user set explicitly, in case it was a default value it will convert it to more appropriate array. There is nothing to check.

Comment (ekohl) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: My interpretation of Upgrade Warnings is that it's things the user should be aware of when upgrading. Changing behavior is a great example of that and this PR does change behavior. Most users won't care, but those who do care about details, typically appreciate it.

@tbrisker does my definition of Upgrade Warnings match yours?

**Review by ares — APPROVED** [review]
Tested, took me a long time before I realize I need to stop spring before actually testing the migration. But then it migrated the setting properly. Verified by using dns_lookup macro, thanks @lzap, merging!

File: db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb
Comment (ares) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: I agree we should document this, saying something like "if you changed dns_conflict_timeout" setting before, you may want to revisit and modify the new value to something more appropriate for your environment, explaining the array format"

Comment (ekohl) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: Also that it has a new name to better cover what it's actually used for. That way users know to look for the right setting.

Comment (lzap) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: Sure, where do we put Upgrade Warnings @tbrisker ? Is this like normal Upgrade Notes we have in the docs or are we talking about something different?

Comment (ekohl) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: The above link (https://github.com/theforeman/theforeman.org/blob/gh-pages/_includes/manuals/nightly/1.2_release_notes.md#upgrade-warnings) is the one I was thinking about, so indeed the regular Upgrade Notes.

Comment (lzap) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: Found it here: https://theforeman.org/manuals/1.22/index.html#Upgradewarnings

Created: https://community.theforeman.org/t/release-notes-and-headline-features-for-1-24/14877

Comment (ekohl) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:32]: We have nightly for staging the 1.24 release. 1.23 has already been branched into its own files.

File: db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb
Comment (lzap) [inline db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:29]: I think this can cause the issue reported in https://projects.theforeman.org/issues/27585

If the old setting was (for any reason) `nil`, this will leave the new setting also `nil` but it's expected now to be an array.

Comment (tbrisker) [reply db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb:29]: This is indeed the bug, but not exactly for that reason: 
This is only checking if both settings exist in the database - not if their value is `nil`. Then, since we override `value` and `value=` the following code does:
```ruby
if old_setting.value != 3 # -> true, since old_setting.value == nil
  new_setting.value = [nil]
```
And dig in to `value=` we get:
```ruby
  def value=([nil])
    v = v.to_yaml unless v.nil?  # v = "---\n- \n" since [nil].nil? == false
    write_attribute :value, v # we now have "---\n- \n" in the database
  end
```

Which means that our fix for this was wrong as well - we assumed the issue was that `nil` was saved, not `[nil]`.

**Pull request merged**
